src/ir_def/check_ir.ml

```diff
@@ -473,6 +473,19 @@ let rec check_exp env (exp:Ir.exp) : unit =
       in
       typ exp2 <: T.nat;
       T.as_immut t2 <: t
+    | GetPastArrayOffset _, [exp1] ->
+      let t1 = T.promote (typ exp1) in
+      ignore
+        (try T.as_array_sub t1 with
+         | Invalid_argument _ ->
+           error env exp1.at "expected array type, but expression produces type\n  %s"
+             (T.string_of_typ_expand t1));
+      T.nat <: t
+    | NextArrayOffset _, [exp1] ->
+      typ exp1 <: T.nat
+    | ValidArrayOffset, [exp1; exp2] ->
+      typ exp1 <: T.nat;
+      typ exp2 <: T.nat
```

Comment:
So, according to your comment, sometime the Nat is actually an Int? So is the ir-checker actually fibbing? If so, that makes me little nervous....

Reply:
Here it is always `Nat`. It changes meaning (slightly) in the Wasm realm.

Reply:
```suggestion
    | GetPastArrayOffset _, [exp1] ->
      let t1 = T.promote (typ exp1) in
      ignore
        (try T.as_array_sub t1 with
         | Invalid_argument _ ->
           error env exp1.at "expected array type, but expression produces type\n  %s"
             (T.string_of_typ_expand t1));
      T.nat <: t
    | NextArrayOffset _, [exp1] ->
      typ exp1 <: T.nat
    | ValidArrayOffset, [exp1; exp2] ->
      typ exp1 <: T.nat;
      typ exp2 <: T.nat;
      T.bool <: t
```
Doesn't this need to return a Boolean?

Reply:
Yes, indeed! Thanks for catching!

Reply:
Line 485 has the same deficiency. Fixing too!

Reply:
Hah, you spotted it: f77424577